Grbl_Esp32/src/Pins/GPIOPinDetail.cpp

```diff
@@ -0,0 +1,174 @@
+            } else if (opt.is("pd")) {
+                _attributes = _attributes | PinAttributes::PullDown;
+            } else if (opt.is("low")) {
+                _attributes = _attributes | PinAttributes::ActiveLow;
+            } else if (opt.is("high")) {
+                // Default: Active HIGH.
+            } else {
+                Assert(false, "Bad GPIO option passed to pin: %s", opt());
+            }
+        }
+
+        // Update the R/W mask for ActiveLow setting
+        if (_attributes.has(PinAttributes::ActiveLow)) {
+            _readWriteMask = HIGH;
+        } else {
+            _readWriteMask = LOW;
+        }
+    }
+
+    PinCapabilities GPIOPinDetail::capabilities() const { return _capabilities; }
+
+    void GPIOPinDetail::write(int high) {
+        Assert(_attributes.has(PinAttributes::Output), "Pin has no output attribute defined. Cannot write to it.");
+        int value = _readWriteMask ^ high;
+        __digitalWrite(_index, value);
+    }
+    int GPIOPinDetail::read() {
+        auto raw = __digitalRead(_index);
+        return raw ^ _readWriteMask;
+    }
+
+    void GPIOPinDetail::setAttr(PinAttributes value) {
+        // These two assertions will fail if we do them for index 1/3 (Serial uart). This is because
+        // they are initialized by HardwareSerial well before we start our main operations. Best to
+        // just ignore them for now, and figure this out later. TODO FIXME!
+
+        // Check the attributes first:
+        Assert(value.validateWith(this->_capabilities) || _index == 1 || _index == 3,
+               "The requested attributes don't match the pin capabilities");
+        Assert(!_attributes.conflictsWith(value) || _index == 1 || _index == 3,
+               "Attributes on this pin have been set before, and there's a conflict.");
+
+        _attributes = value;
+
+        // Handle attributes:
+        uint8_t pinModeValue = 0;
+
+        if (value.has(PinAttributes::Input)) {
+            pinModeValue |= INPUT;
+        } else if (value.has(PinAttributes::Output)) {
+            pinModeValue |= OUTPUT;
+        }
+
+        // PU/PD should be specified by the user. Code has nothing to do with them:
+        if (_attributes.has(PinAttributes::PullUp)) {
+            pinModeValue |= PULLUP;
+        } else if (_attributes.has(PinAttributes::PullDown)) {
+            pinModeValue |= PULLDOWN;
+        }
+
+        // If the pin is ActiveLow, we should take that into account here:
+        if (value.has(PinAttributes::InitialOn)) {
+            __digitalWrite(_index, HIGH ^ _readWriteMask);
+        } else {
+            __digitalWrite(_index, LOW ^ _readWriteMask);
+        }
+
+        __pinMode(_index, pinModeValue);
+    }
+
+    void GPIOPinDetail::attachInterrupt(void (*callback)(void*), void* arg, int mode) {
+        Assert(_attributes.has(PinAttributes::ISR), "Pin has no ISR attribute defined. Cannot bind ISR.");
```

Comment:
My mistake, it is the Assert that is causing the crash. Message does not print. Just crashes. Comment out the assert and it runs fine.

Reply:
Here is the fix that works for me.
    void GPIOPinDetail::attachInterrupt(void (*callback)(void*), void* arg, int mode) {
        Assert(_capabilities.has(PinCapabilities::ISR), "Pin has no ISR capability defined. Cannot bind ISR.");
        ::attachInterruptArg(_index, callback, arg, mode);
    }

    void GPIOPinDetail::detachInterrupt() {
        Assert(_capabilities.has(PinCapabilities::ISR), "Pin has no ISR capability defined. Cannot unbind ISR.");
        ::detachInterrupt(_index);
    }

Reply:
Nicely caught what's going on here.

Yes, the assert throws an exception. That being said, I was actually surprised that it led to a crash. Assertions should not allow the program to continue (unless...), but they should _always_ provide the developer with all the information he needs to fix the issue at hand. Basically the 'assert' is like 'if this happens, things are terribly wrong in our code'. 

After inspection, I noticed that not all exceptions are being caught and added a catch in the grbl_init. Added some code for that.

.

We should fix this by getting our definitions straight. I'm also writing this down for my own reference. Basically it works as follows:

(1) PinCapabilities define hardware capabilities. This is just a given, and always defines capabilities. This is just the way it is. [hardware]
(2) PinOptions define attributes of the pin as requested by the user. PinOptions here usually defines input, output, pu, pd, and active low. These are checked against the capabilities, and if there is no conflict, are stored as attributes in the pin class. [user]
(3) Code uses a pin class, see below [code]

Code must always do the following things, in order:
(a) Set the requested attributes on the pin. This basically validates the request with [user] and [hardware], and if successful performs a 'setMode' or whatever is necessary to set it up. So, a [hardware] input pin cannot be bound as 'output', a [user] input pin cannot be bound as 'output' and a non-ISR [hardware] pin cannot be bound as ISR pin.
(b) Performs operations on the pin. This validates the operation with the requested attributes from (a) and does the thing.

.

What happens here is that ISR is bound (b), without ISR being initialized (a). So, I think we should not fix this by changing capabilities to attributes in the assert, but by fixing the setAttr call earlier -- that should include ISR. I'm quickly going through the code now to fix these where attach/detach are being called, that should remove some of the bugs.